internal/graph/iterator/iterator_stream.go

```diff
@@ -0,0 +1,203 @@
+package iterator
+
+import (
+	"context"
+	"fmt"
+	"slices"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+type Msg struct {
+	Iter storage.TupleKeyIterator
+	Err  error
+}
+
+type Stream struct {
+	idx    int
+	buffer storage.TupleKeyIterator
+	done   bool // done is set when the buffer is Done and all the source are exhausted
+	source chan *Msg
+}
+
+func NewStream(idx int, source chan *Msg) *Stream {
+	return &Stream{
+		idx:    idx,
+		source: source,
+	}
+}
+
+func (s *Stream) Idx() int {
+	return s.idx
+}
+
+// Head returns the first item in the buffer.  If the Head is done or
+// cancelled, it will stop the buffer and set the buffer to nil.
+func (s *Stream) Head(ctx context.Context) (*openfgav1.TupleKey, error) {
+	if s.buffer == nil {
+		return nil, storage.ErrIteratorDone
+	}
+	t, err := s.buffer.Head(ctx)
+	if err != nil {
+		if storage.IterIsDoneOrCancelled(err) {
+			s.buffer.Stop()
+			s.buffer = nil
+		}
+		return nil, err
+	}
+	return t, nil
+}
+
+func (s *Stream) Next(ctx context.Context) (*openfgav1.TupleKey, error) {
+	if s.buffer == nil {
+		return nil, storage.ErrIteratorDone
+	}
+	t, err := s.buffer.Next(ctx)
+	if err != nil {
+		if storage.IterIsDoneOrCancelled(err) {
+			s.buffer.Stop()
+			s.buffer = nil
+		}
+		return nil, err
+	}
+	return t, nil
+}
+
+func (s *Stream) Stop() {
+	if s.buffer != nil {
+		s.buffer.Stop()
+	}
+	for msg := range s.source {
+		if msg.Iter != nil {
+			msg.Iter.Stop()
+		}
+	}
+}
+
+// SkipToTargetObject moves the buffer until the buffer's head object is >= target object.
+// If the buffer is drained and no more items, it will set to stop and buffer will be nil.
+func (s *Stream) SkipToTargetObject(ctx context.Context, target string) error {
+	if !tuple.IsValidObject(target) {
+		return fmt.Errorf("invalid target object: %s", target)
+	}
+
+	t, err := s.Head(ctx)
+	if err != nil {
+		if storage.IterIsDoneOrCancelled(err) {
+			return nil
+		}
+		return err
+	}
+	tmpKey := t.GetObject()
+	for tmpKey < target {
+		_, _ = s.Next(ctx)
+		t, err = s.Head(ctx)
+		if err != nil {
+			if storage.IterIsDoneOrCancelled(err) {
+				break
+			}
+			return err
+		}
+		tmpKey = t.GetObject()
+	}
+	return nil
+}
+
+// Drain all item in the stream's buffer and return these items.
+func (s *Stream) Drain(ctx context.Context) ([]*openfgav1.TupleKey, error) {
+	var batch []*openfgav1.TupleKey
+	for {
+		t, err := s.Next(ctx)
+		if err != nil {
+			if storage.IterIsDoneOrCancelled(err) {
+				break
+			}
+			return nil, err
+		}
+		batch = append(batch, t)
+	}
+	return batch, nil
+}
+
+func (s *Stream) fetchSource(ctx context.Context) error {
+	if s.buffer != nil || s.done {
+		// no need to poll further
+		return nil
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case i, ok := <-s.source:
+		if !ok {
+			s.done = true
+			break
+		}
+		if i.Err != nil {
+			return i.Err
+		}
+		s.buffer = i.Iter
```

Comment:
Could it happen that before line 142, `s.buffer` was already not nil?  If not, it seems like there is an unwritten pre-condition that this function has: that the consumer of the stream must have already called Head or Next until `storage.ErrIteratorDone`.